### PR TITLE
Refactor NetworkAgentRegistry

### LINKED ISSUE #432: Refactor NetworkAgentRegistry
# For New Contributors
[How to contribute](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/wiki/Getting-Started-as-a-Contributor)

## Description

`NetworkAgentRegistry` requires a new dictionary or hashset for player agents. This will be used by the method `IsNetworkAgent` and then by `AgentInteractionPatch`.

Currently the way this works is any controlled agent even not directly player controlled returns true for `IsNetworkAgent`.

Create another extension method called `IsPlayerAgent` that returns true if the given agent is directly controlled by any player.

## Intended Design

N/A

## Location

Branch from taverns-vertical slice and merge back when complete.

## Related Issues

N/A

## Requirements

N/A

## Additional information

N/A

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit tests exist for every method that does not require the game to be ran.

### LINKED ISSUE #485: Convert AgentRegistry Dictionaries to ConcurrentDictionaries
# For New Contributors
[How to contribute](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/wiki/Getting-Started-as-a-Contributor)

## Description

We want to add concurrent functionality to the `AgentRegistry` class. Update the current Dictionaries to ConcurrentDictionaries.

Also create a test for every method in the `AgentRegistry` class

## Intended Design

N/A

## Location

Create a new branch based from `Arena-Vertical-Slice`

Tests should exist under `MissionTests` project in a folder named `Network`. Name the new test file `AgentRegistryTests`

## Related Issues

Blocked by
- #432 

## Requirements

N/A

## Additional information

N/A

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit tests exist for every method that does not require the game to be ran.

### PR DESCRIPTION
**Pull request by Ciel1996**
- Added IsPlayerAgent extension-method using this dictionary
- Added methods to register PlayerAgents and NetworkPlayerAgents
- Adjusted AgentInteractionPatches to use IsPlayerAgent instead of IsNetworkAgent

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] All new classes have class-level documentation comments, if there are any at all
- [ ] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:

## What is the current behaviour?
Currently only all controlled agents of players are stored.

## What is the new behaviour?
Resolves #432
- Added new Dictionary and methods to store PlayerAgents.
- Added new extension method `IsPlayerAgent`
- Adjusted AgentInteractionPatches to use `IsPlayerAgent` instead of `IsNetworkAgent`
- 
PlayerAgents can be registered in the NetworkAgentRegistry both locally and remotely.
`IsPlayerAgent` can be used to check whether an agent is directly controlled by a player.
Behaviour of `IsNetworkAgent` remains unchanged.

Resolves: #485 
- Replaced Dictionary with ConcurrentDictionary
- Added tests to cover functionality

## Other information
Used VS feature to detect and automatically remove unused using directives.

### CODE REVIEW

File: source/Missions/Services/Network/NetworkAgentRegistry.cs [hunk 1]
```diff
@@ -64,11 +71,33 @@ private void Handle_AgentDeleted(MessagePayload<AgentDeleted> payload)
...
+                    return false;
+                }
+            });
+        }
 
         public bool RegisterControlledAgent(Guid agentId, Agent agent)
         {
-            if (m_AgentToId.ContainsKey(agent)) return false;
-            if (m_ControlledAgents.ContainsKey(agentId)) return false;
+            if (m_AgentToId.ContainsKey(agent) ||
```
Comment (garrettluskey) [inline source/Missions/Services/Network/NetworkAgentRegistry.cs]: It's slightly better to keep this as 2 if statements just for constancy as well as readability

File: source/Missions/Services/Network/NetworkAgentRegistry.cs [hunk 2]
```diff
@@ -80,18 +109,50 @@ public bool RegisterNetworkControlledAgent(NetPeer peer, Guid agentId, Agent age
         {
             if (m_OtherAgents.TryGetValue(peer, out AgentGroupController controller))
             {
-                if (controller.Contains(agent)) return false;
-                if (controller.Contains(agentId)) return false;
+                if (controller.Contains(agent) || controller.Contains(agentId)) return false;
```
Comment (garrettluskey) [inline source/Missions/Services/Network/NetworkAgentRegistry.cs]: It's slightly better to keep this as 2 if statements just for constancy as well as readability

File: source/Missions/Services/Network/NetworkAgentRegistry.cs [hunk 3]
```diff
@@ -80,18 +109,50 @@ public bool RegisterNetworkControlledAgent(NetPeer peer, Guid agentId, Agent age
...
+            m_AgentToId.Add(agent, agentId);
+
+            return true;
+        }
+
+        public bool RegisterNetworkPlayerAgent(NetPeer peer, Guid agentId, Agent agent)
+        {
+            if (m_OtherAgents.TryGetValue(peer, out AgentGroupController controller))
+            {
+                if (controller.Contains(agent) || controller.Contains(agentId)) return false;
```
Comment (garrettluskey) [inline source/Missions/Services/Network/NetworkAgentRegistry.cs]: It's slightly better to keep this as 2 if statements just for constancy as well as readability

File: source/Missions/Services/Network/NetworkAgentRegistry.cs [hunk 4]
```diff
@@ -64,11 +71,33 @@ private void Handle_AgentDeleted(MessagePayload<AgentDeleted> payload)
...
+            {
+                Agent agent = group.RemoveAgent(agentId);
+                if (agent != null)
+                {
+                    var result = m_AgentToId.Remove(agent);
+
+                    if (m_PlayerAgents.ContainsKey(agentId))
+                    {
+                        result &= m_PlayerAgents.Remove(agentId);
+                    }
```
Comment (garrettluskey) [inline source/Missions/Services/Network/NetworkAgentRegistry.cs]: m_PlayerAgents.Remove(agentId) in this if will always return true, so we don't need to add it to the result.

We won't care about the result of `m_PlayerAgents.Remove(agentId)` due to this method's behavior being to primarily remove NetworkControlledAgent
```suggestion
                    RemovePlayerAgent(agentId);
```

File: source/Missions/Services/Network/NetworkAgentRegistry.cs [hunk 5]
```diff
@@ -106,28 +167,32 @@ public bool RemoveControlledAgent(Guid agentId)
...
-                }
-                else
-                {
-                    return false;
-                }
-            });
+                return m_AgentToId.Remove(agent);
+            }
+
+            return false;
```
Comment (garrettluskey) [inline source/Missions/Services/Network/NetworkAgentRegistry.cs]: Change this method to only be `return m_PlayerAgents.Remove(agentId)` all the other stuff is not needed

File: source/Missions/Services/Network/NetworkAgentRegistry.cs [hunk 6]
```diff
@@ -106,28 +167,32 @@ public bool RemoveControlledAgent(Guid agentId)
...
             {
-                result &= controller.ControlledAgents.All(kvp => m_AgentToId.Remove(kvp.Value));
+                result &= controller.ControlledAgents.All(kvp => 
+                {
+                    var innerResult = m_AgentToId.Remove(kvp.Value);
+
+                    if (kvp.Value.IsPlayerAgent())
+                    {
+                        innerResult &= m_PlayerAgents.Remove(kvp.Key);
+                    }
```
Comment (garrettluskey) [inline source/Missions/Services/Network/NetworkAgentRegistry.cs]:
```suggestion
                    RemovePlayerAgent(kvp.Key);
```

File: source/Missions/Services/Network/NetworkAgentRegistry.cs [hunk 7]
```diff
@@ -80,18 +109,50 @@ public bool RegisterNetworkControlledAgent(NetPeer peer, Guid agentId, Agent age
...
+                m_AgentToId.Add(agent, agentId);
+                m_OtherAgents.Add(peer, newGroupController);
+            }
+
+            return true;
+        }
+
+        public bool RegisterPlayerAgent(Guid agentId, Agent agent)
+        {
+            if (m_AgentToId.ContainsKey(agent) ||
```
Comment (garrettluskey) [inline source/Missions/Services/Network/NetworkAgentRegistry.cs]: It's slightly better to keep this as 2 if statements just for constancy as well as readability

File: source/Missions/Services/Network/NetworkAgentRegistry.cs [hunk 8]
```diff
@@ -80,18 +109,50 @@ public bool RegisterNetworkControlledAgent(NetPeer peer, Guid agentId, Agent age
...
+                m_PlayerAgents.ContainsKey(agentId)) return false;
+
+            m_PlayerAgents.Add(agentId, agent);
+            m_AgentToId.Add(agent, agentId);
```
Comment (garrettluskey) [inline source/Missions/Services/Network/NetworkAgentRegistry.cs]:
```suggestion
            RegisterControlledAgent(agent, agentId);
```

File: source/Missions/Services/Network/NetworkAgentRegistry.cs [hunk 9]
```diff
@@ -80,18 +109,50 @@ public bool RegisterNetworkControlledAgent(NetPeer peer, Guid agentId, Agent age
...
+
+        public bool RegisterNetworkPlayerAgent(NetPeer peer, Guid agentId, Agent agent)
+        {
+            if (m_OtherAgents.TryGetValue(peer, out AgentGroupController controller))
+            {
+                if (controller.Contains(agent) || controller.Contains(agentId)) return false;
 
                 controller.AddAgent(agentId, agent);
                 m_AgentToId.Add(agent, agentId);
+                m_PlayerAgents.Add(agentId, agent);
```
Comment (garrettluskey) [inline source/Missions/Services/Network/NetworkAgentRegistry.cs]:
```suggestion
```

File: source/Missions/Services/Network/NetworkAgentRegistry.cs [hunk 10]
```diff
@@ -80,18 +109,50 @@ public bool RegisterNetworkControlledAgent(NetPeer peer, Guid agentId, Agent age
...
             }
             else
             {
                 AgentGroupController newGroupController = new AgentGroupController();
                 newGroupController.AddAgent(agentId, agent);
                 m_AgentToId.Add(agent, agentId);
                 m_OtherAgents.Add(peer, newGroupController);
+                m_PlayerAgents.Add(agentId, agent);
```
Comment (garrettluskey) [inline source/Missions/Services/Network/NetworkAgentRegistry.cs]:
```suggestion
```

File: source/Missions/Services/Agents/Extensions/AgentExtensions.cs [hunk 11]
```diff
@@ -1,13 +1,19 @@
...
     {
-        public static bool IsNetworkAgent(this Agent agent)
+        public static bool IsNetworkAgent(this Agent agent) 
         {
             return NetworkAgentRegistry.Instance.AgentToId.ContainsKey(agent);
         }
+
+        public static bool IsPlayerAgent(this Agent agent)
+        {
+            return NetworkAgentRegistry.Instance.PlayerAgents.Any(kvp => kvp.Value == agent);
```
Comment (garrettluskey) [inline source/Missions/Services/Agents/Extensions/AgentExtensions.cs]: O(1) is possible here
```suggestion
    if (NetworkAgentRegistry.Instance.AgentToId.TryGetValue(agent, out Guid id)
    {
            return NetworkAgentRegistry.Instance.PlayerAgents.Contains(id);
    }
    
    return false;
```

**Review by garrettluskey — CHANGES_REQUESTED** [review]
Nice updates. This problem was hard to explain but even still very close to what I wanted. Thanks for the help!

File: source/Missions/Services/Network/NetworkAgentRegistry.cs (hunk 9 above)
Comment (Ciel1996) [reply source/Missions/Services/Network/NetworkAgentRegistry.cs]: that would make the whole PlayerAgents dictionary obsolete, wouldn't it?

File: source/Missions/Services/Network/NetworkAgentRegistry.cs (hunk 10 above)
Comment (Ciel1996) [reply source/Missions/Services/Network/NetworkAgentRegistry.cs]: that would make the whole PlayerAgents dictionary obsolete, wouldn't it?

File: source/Missions/Services/Network/NetworkAgentRegistry.cs [hunk 12]
```diff
@@ -80,18 +109,50 @@ public bool RegisterNetworkControlledAgent(NetPeer peer, Guid agentId, Agent age
...
+            if (m_AgentToId.ContainsKey(agent) ||
+                m_PlayerAgents.ContainsKey(agentId)) return false;
+
+            m_PlayerAgents.Add(agentId, agent);
+            m_AgentToId.Add(agent, agentId);
+
+            return true;
+        }
+
+        public bool RegisterNetworkPlayerAgent(NetPeer peer, Guid agentId, Agent agent)
```
Comment (garrettluskey) [inline source/Missions/Services/Network/NetworkAgentRegistry.cs:136]: Change this method to 
```
if(RegisterNetworkControlledAgent(peer, agentId, agent) == false) return false;
if(m_PlayerAgents.ContainsKey(agentId)) return false;

m_PlayerAgents.Add(agentId, agent);
return true;
```

**Review by garrettluskey — CHANGES_REQUESTED** [review]

File: source/Missions/Services/Network/NetworkAgentRegistry.cs (hunk 12 above)
Comment (Ciel1996) [reply source/Missions/Services/Network/NetworkAgentRegistry.cs:136]: I also refactored `RegisterPlayerAgent` in a similar way.

**Comment by Ciel1996** [comment]
Blocker: Will also address #485

**Comment by garrettluskey** [comment]
Changes look good. Feel free to start on #485

**Comment by Ciel1996** [comment]
@garrettluskey made the changes for #485, please have a look again

**Review by garrettluskey — APPROVED** [review]
Nice work

**Pull request merged**